pytket_qir/converter.py

```diff
@@ -0,0 +1,371 @@
+from collections import OrderedDict
+from dataclasses import dataclass
+import os
+from typing import Optional, cast, List, Union
+
+from pytket import Circuit  # type: ignore
+from pytket.circuit import CircBox  # type: ignore
+from pytket.circuit.logic_exp import BitNot, if_bit  # type: ignore
+from pytket.wasm import WasmFileHandler  # type: ignore
+
+from pyqir.parser import (  # type: ignore
+    QirBlock,
+    QirBrTerminator,
+    QirCondBrTerminator,
+    QirLocalOperand,
+    QirModule,
+    QirRetTerminator,
+)
+from pyqir.generator import ir_to_bitcode, types  # type: ignore
+
+from pytket_qir.gatesets.base import CustomGateSet
+from pytket_qir.parser import QirParser
+from pytket_qir.utils import ConverterException
+
+
+@dataclass
+class Block:
+    name: str
+    succs: List
+    preds: List
+    composition: List
+    visited: bool = False
+
+
+def topological_sort(digraph: dict) -> dict:
```

Comment:
Should we use a third-party library like `networkx` for graph manipulation? Then we'd be able to use the build-in algorithms, such as top sort and cyclicity checking.

Reply:
Could be. One thing that can be interesting is if `networkx` allows for a user to define custom graph rewrites more easily. Then, one can come up with other and better CFG optimisations (than my naive one about block contractions) and apply them. At least, it could provide a better framework for doing this. However, after a quick glance at the doc, it doesn't seem to provide that functionality. Later, in case of QIR programs that exhibit back edges, that can be interesting if loops aren't fully unrolled.

Reply:
I have thought about replacing this with something else as well, but I thought that it might be better in the long term to not have that much optimisation going on in the mainly conversion module. I think we should try to parse this in the long term into something that allows us to optimise this on the c++ level?

Reply:
Could be. Although it could also fit into the TKET2 AST Seyon mentioned elsewhere. As far as I've understood, TKET2 is designed to handle conditionals and classical arithmetic in a better way than what TKET1 currently does. I am not sure how much the AST overlaps/captures the CFG structure but if it does to some extend, it would make sense that optimisations happen there.